### DISCUSSION TITLE
par has error:  par.tcl, 29 invalid command name "ln".

### DISCUSSION THREAD

**Discussion by ZhaoHaowenn** [ucb-bar/hammer#728] [Support (Q&A)] [answered]
Hi 
When I run `make par tutorial=ky130-openroad`,  "ln" command in par.tcl isn't valid and par failed. Here are the error messages:

create_clock clock_clock -name clock_clock -period 30.0
set_clock_uncertainty 1.0 [get_clocks clock_clock]
set_clock_groups -asynchronous  -group { clock_clock }
set_load 1.0 [all_outputs]
write_db pre_floorplan_design
Error: par.tcl, 29 invalid command name "ln"
Action par config output written to output.json

And some commands related to this bug in par.tcl is:
....
puts "write_db pre_floorplan_design"
write_db pre_floorplan_design
ln -sfn pre_floorplan_design latest

I think there is a missing "puts" instruction before the “ln” like other command in tcl.  But I can't manually add it because rerunning par will regenerate this tcl file. How to deal with it?

**Comment by harrisonliew** [comment, marked as answer]
There was a bug, which was fixed in #707. There should be an `exec` before the `ln` to indicate that it should be executed in the system shell instead of OpenROAD's shell.
https://github.com/ucb-bar/hammer/blob/c85bd2a093e105bece4b10bfaf1438825499ce0d/hammer/par/openroad/__init__.py#L109

If you're willing to wait until next week, these fixes will make their way into Chipyard.

**Reply by ZhaoHaowenn** [reply]
Thanks for explanation. I have fixed it. And I will wait for next week.

**Comment by ZhaoHaowenn** [comment]
Hi
I found that due to the openroad update, the flags of some openroad command in `hammer/hammer/par/openroad/__init__.py` are no longer supported.
par issue:
> (hammer) pdngen -verbose /root/chipyard/vlsi/build-sky130-openroad/chipyard.TestHarness.TinyRocketConfig-ChipTop/par-rundir/power_straps.pdn
[ERROR STA-0406] pdngen positional arguments not supported.
Error: par.tcl, 74 STA-0406

in `hammer/hammer/par/openroad/__init__.py`:
https://github.com/ucb-bar/hammer/blob/92e98c1a2c46f207cdbf6aeb134c3c6d03c79fd9/hammer/par/openroad/__init__.py#L706

par issue:
> (hammer) detailed_route -guide /root/chipyard/vlsi/build-sky130-openroad/chipyard.TestHarness.TinyRocketConfig-ChipTop/par-rundir/ChipTop.route_guide
[ERROR STA-0402] detailed_route -guide is not a known keyword or flag.
Error: par.tcl, 316 STA-0402

in `hammer/hammer/par/openroad/__init__.py`, the "guide" flag is no longer supported:
https://github.com/ucb-bar/hammer/blob/92e98c1a2c46f207cdbf6aeb134c3c6d03c79fd9/hammer/par/openroad/__init__.py#L985

I want to know which version of OpenRoad that Hammer supports or if these will be fixed soon. Thanks again for the reply.

**Reply by harrisonliew** [reply]
They will be fixed soon: https://github.com/ucb-bar/hammer/pull/714/files#diff-a7c0571e0f65aab63321fa7e7893fd79677193ae36ed7b2dd39a9579730d306b

We will be supporting the latest version that is published to conda.

**Reply by ZhaoHaowenn** [reply]
That's great! thanks.
